platform/viewer/src/connectedComponents/ViewerRetrieveStudyData.js

```diff
@@ -1,77 +1,118 @@
+    ).then(
+      studies => {
+        if (this.isReady) {
+          this.setStudies(studies);
+        }
+      },
+      error => {
+        if (this.isReady) {
+          this.setState({ error: true });
+        }
+        log.error(error);
+      }
+    );
+  }
 
+  setStudies(givenStudies) {
+    if (Array.isArray(givenStudies) && givenStudies.length > 0) {
+      const sopClassHandlerModules =
+        extensionManager.modules['sopClassHandlerModule'];
       // Map studies to new format, update metadata manager?
-      const updatedStudies = studies.map(study => {
+      const studies = givenStudies.map(study => {
         const studyMetadata = new OHIFStudyMetadata(
           study,
           study.studyInstanceUid
         );
-        const sopClassHandlerModules =
-          extensionManager.modules["sopClassHandlerModule"];
-
-        study.displaySets =
-          study.displaySets ||
-          studyMetadata.createDisplaySets(sopClassHandlerModules);
+        if (!study.displaySets) {
+          study.displaySets = studyMetadata.createDisplaySets(
+            sopClassHandlerModules
+          );
+        }
         studyMetadata.setDisplaySets(study.displaySets);
-
         // Updates WADO-RS metaDataManager
         updateMetaDataManager(study);
-
         studyMetadataManager.add(studyMetadata);
-
+        // Attempt to load remaning series if any
+        this._attemptToLoadRemainingSeries(study, studyMetadata);
         return study;
       });
+      this.setState({ studies });
+    }
+  }
 
-      this.setState({
-        studies: updatedStudies
+  _attemptToLoadRemainingSeries(study, studyMetadata) {
+    const { seriesLoadQueue } = study;
+    if (seriesLoadQueue) {
+      const view = this;
+      study.subscribe('seriesAdded', function seriesAddedHandler(series) {
+        if (view.isReady) {
+          const sopClassHandlerModules =
+            extensionManager.modules['sopClassHandlerModule'];
+          const seriesMetadata = new OHIFSeriesMetadata(series, study);
+          studyMetadata.addSeries(seriesMetadata);
+          studyMetadata.createAndAddDisplaySetsForSeries(
+            sopClassHandlerModules,
+            seriesMetadata
+          );
+          study.displaySets = studyMetadata.getDisplaySets();
+          updateMetaDataManager(study, series.seriesInstanceUid);
+          view.setState(function(state) {
+            return { studies: state.studies.slice() };
+          });
+        } else {
+          study.unsubscribe('seriesAdded', seriesAddedHandler);
+        }
       });
-    } catch (err) {
-      this.setState({
-        error: true
-      });
-
-      // TODO: Handle gracefully instead of throwing?
-      throw new Error(err);
+      while (seriesLoadQueue.size() > 0) {
+        seriesLoadQueue.dequeue();
+      }
     }
   }
 
+  componentDidMount() {
+    // TODO: CLEAR THIS SOMEWHERE ELSE
+    studyMetadataManager.purge();
+    this.isReady = true;
+    this.triggerDataLoad();
```

Comment:
Wait. So `isReady` is always true when `triggerDataLoad()` is called? This would only _not_ be true if the component was destroyed before the query resolved.

- Request the data, and create a cancellation token
- When the request resolves, clear the token
- In `willUnmount`, if the token is set, use it to cancel the request

Reply:
Yes. But I already removed the `this.isReady` flag and added a `this.subscriptions` list which will take care of getting rid of the registered handlers if the component is unmounted.